### PR TITLE
Drawer, gridview, pageview and tests added with multiple lists of movies connected to TMDB

### PR DESCRIPTION
**Pull request by evibart**
Movie Lists implemented: Top viewed, Popular, Now playing and Upcoming.
Implemented: widget and unit testing.
https://github.com/evibart/TheMovieDBApp/assets/142030257/c3ac5dfd-3d6c-4adf-8700-ca3e8cfe52be

### CODE REVIEW

File: lib/src/data/datasource/remote/api_service.dart [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+          Constants.apiBaseUrl +
+              '/' +
+              Constants.apiVersion +
+              '/' +
+              Constants.apiMoviePath +
+              '/' +
+              Constants.apiPopularMoviesPath +
+              '?' +
+              Constants.apiKey,
+        ),
```
Comment (LeonardoMolleker) [inline lib/src/data/datasource/remote/api_service.dart]: i recommend use string interpolation for a clearer code

File: lib/src/data/datasource/remote/api_service.dart [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        Uri.parse(
+          Constants.apiBaseUrl +
+              '/' +
+              Constants.apiVersion +
+              '/' +
+              Constants.apiMoviePath +
+              '/' +
+              Constants.apiNowPlaying +
+              '?' +
+              Constants.apiKey,
```
Comment (LeonardoMolleker) [inline lib/src/data/datasource/remote/api_service.dart]: same as above

File: lib/src/data/datasource/remote/api_service.dart [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        Uri.parse(
+          Constants.apiBaseUrl +
+              '/' +
+              Constants.apiVersion +
+              '/' +
+              Constants.apiMoviePath +
+              '/' +
+              Constants.apiTopRated +
+              '?' +
+              Constants.apiKey,
```
Comment (LeonardoMolleker) [inline lib/src/data/datasource/remote/api_service.dart]: same

File: lib/src/data/datasource/remote/api_service.dart [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        Uri.parse(
+          Constants.apiBaseUrl +
+              '/' +
+              Constants.apiVersion +
+              '/' +
+              Constants.apiMoviePath +
+              '/' +
+              Constants.apiUpComingMovies +
+              '?' +
+              Constants.apiKey,
```
Comment (LeonardoMolleker) [inline lib/src/data/datasource/remote/api_service.dart]: same

File: lib/src/data/datasource/remote/api_service.dart [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            '/' +
+            Constants.apiVersion +
+            '/' +
+            Constants.apiGenre +
+            '/' +
+            Constants.apiMoviePath +
+            '/' +
+            Constants.apiList +
+            '?' +
+            Constants.apiKey),
```
Comment (LeonardoMolleker) [inline lib/src/data/datasource/remote/api_service.dart]: same

File: lib/src/data/repository/movie_remote_repository.dart [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                ),
+              )
+              .toList(),
+        );
+      } else {
+        throw Exception('${response.statusCode} ${response.body}');
+      }
+    } catch (e) {
+      return Left(e.toString());
+    }
```
Comment (LeonardoMolleker) [inline lib/src/data/repository/movie_remote_repository.dart]: there is a lot of duplicated code here, find a way to abstract this method, you are doing the same the only variant is the endpoint

File: lib/src/presentation/bloc/bloc_movie_list.dart [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        await upComingMoviesUseCase.execute().fold(
+              (error) => _movieList.sink.addError(error),
+              (movieList) => _movieList.sink.add(movieList),
+            );
+        break;
+      default:
+        await popularMoviesUseCase.execute().fold(
+              (error) => _movieList.sink.addError(error),
+              (movieList) => _movieList.sink.add(movieList),
+            );
```
Comment (LeonardoMolleker) [inline lib/src/presentation/bloc/bloc_movie_list.dart]: duplicated code again, let's finde a wat to abstract this code

File: lib/src/presentation/view/movie_list_home.dart [hunk 8]
```diff
@@ -1,53 +1,99 @@
...
+              title: const Text('Now Playing'),
+              key: Key('NowPlayingDrawer'),
+              onTap: () => _onDrawerItemTap(MovieEndpoint.NOW_PLAYING),
+            ),
+            ListTile(
+              title: const Text('Up Coming'),
+              key: Key('UpComingDrawer'),
+              onTap: () => _onDrawerItemTap(MovieEndpoint.UP_COMING),
+            ),
+          ],
```
Comment (LeonardoMolleker) [inline lib/src/presentation/view/movie_list_home.dart]: constants

File: lib/my_app.dart [hunk 9]
```diff
@@ -15,7 +28,18 @@ class MyApp extends StatelessWidget {
       ),
       initialRoute: MovieListHome.routeName,
       routes: {
-        MovieListHome.routeName: (context) => const MovieListHome(),
+        MovieListHome.routeName: (context) => MovieListHome(
+              blocMovieList: BlocMovieList(
```
Comment (FranciscoGaraventa) [inline lib/my_app.dart]: Why you call it BlocMovieList, the bloc class never has a reference of a ui structure

File: lib/src/data/datasource/mappers/genre_mapper.dart [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import '../../../domain/entity/genre_entity.dart';
+import '../../model/genre_data.dart';
+
+class GenreMapper {
+  static Genre fromDataToEntity(GenreData genre) {
+    return Genre(
+      id: genre.id,
+      name: genre.name,
+    );
+  }
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/mappers/genre_mapper.dart]: You can define a Genre Entity like this on domain layer
```
class GenreEntity {
  GenreEntity{
    required this.id,
    required this.name
  });

  final int id;
  final String name;
}
```

Comment (FranciscoGaraventa) [reply lib/src/data/datasource/mappers/genre_mapper.dart]: on data layer you are going to have a GenreModel like this
```
class GenreModel extends GenreEntity {
  GenreModel({
    required int id,
    required String name,
  }) : super(
          id: id,
          name: name,
        );

  factory GenreModel.fromJson(Map<String, dynamic> json) => GenreModel(...);
}
```

Comment (FranciscoGaraventa) [reply lib/src/data/datasource/mappers/genre_mapper.dart]: If you notice, on domain we define the structure that we need to use and presentation is going to use, but on data layer we know how to build it

File: lib/src/data/datasource/mappers/movie_mapper.dart [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+      poster: movieData.poster,
+      backdrop: movieData.backdrop,
+      originalLanguage: movieData.originalLanguage,
+      popularity: movieData.popularity,
+      id: movieData.id,
+      hasVideo: movieData.hasVideo,
+      voteCount: movieData.voteCount,
+      isAdult: movieData.isAdult,
+    );
+  }
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/mappers/movie_mapper.dart]: Same as the comment for the genres mapper

File: lib/src/data/repository/genre_remote_repository.dart [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+class GenreRemoteRepository implements IGenreRepository {
+  final ApiService apiService;
+
+  GenreRemoteRepository({required this.apiService});
+
+  @override
+  EitherErrorManager<List<Genre>> loadGenres() async {
+    try {
+      final response = await apiService.getGenres();
+      if (response.statusCode == Constants.statusCodeOk) {
```
Comment (FranciscoGaraventa) [inline lib/src/data/repository/genre_remote_repository.dart]: Instead to add a const to 200, you can use HttpStatus.OK

File: lib/src/data/repository/movie_remote_repository.dart [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+  MovieRemoteRepository({
+    required this.apiService,
+  });
+
+  @override
+  EitherErrorManager<List<Movie>> loadPopularMovies() async {
+    try {
+      final response = await apiService.getPopularMovies();
+      if (response.statusCode == Constants.statusCodeOk) {
```
Comment (FranciscoGaraventa) [inline lib/src/data/repository/movie_remote_repository.dart]: Same as above

File: lib/src/data/repository/movie_remote_repository.dart [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+import '../../core/util/constants.dart';
+import '../../core/util/either_string_t.dart';
+import '../../domain/entity/movie_entity.dart';
+import '../../domain/repository/movie_repository_i.dart';
+import '../datasource/mappers/movie_mapper.dart';
+import '../datasource/remote/api_service.dart';
+import '../model/movie_data.dart';
+
+class MovieRemoteRepository implements IMovieRepository {
```
Comment (FranciscoGaraventa) [inline lib/src/data/repository/movie_remote_repository.dart:7]: If you can see here, you are replicating the same code 4 times, the only thing that is change in is the endpoint that you use, so instead to made for similar methods, you can use only one and as a parameter you can select which endpoint use

File: lib/src/domain/repository/movie_repository_i.dart [hunk 15]
```diff
@@ -0,0 +1,15 @@
+import '../../core/util/either_string_t.dart';
+import '../entity/movie_entity.dart';
+
+abstract class IMovieRepository<T> {
+  EitherErrorManager<List<Movie>> loadPopularMovies();
```
Comment (FranciscoGaraventa) [inline lib/src/domain/repository/movie_repository_i.dart]: Why is called EitherErrorManager?

File: lib/src/domain/usecase/implementation/up_coming_movies_usecase.dart [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import '../../../core/util/either_string_t.dart';
+import '../../entity/movie_entity.dart';
+import '../../repository/movie_repository_i.dart';
+import '../list_usecase_i.dart';
+
+class UpComingMoviesUseCase implements ListUseCaseI<List<Movie>> {
```
Comment (FranciscoGaraventa) [inline lib/src/domain/usecase/implementation/up_coming_movies_usecase.dart]: Same as i mentioned above, you are creating for separate use case that are exactly the same, the only thing that you change is the repository call, you can keep it like this way, but maybe you can only have one use case, that centralized the 4 repository calls

File: lib/src/domain/usecase/list_usecase_i.dart [hunk 17]
```diff
@@ -0,0 +1,5 @@
+import '../../core/util/either_string_t.dart';
+
+abstract class ListUseCaseI<T> {
```
Comment (FranciscoGaraventa) [inline lib/src/domain/usecase/list_usecase_i.dart]: Instead to ListUseCaseI you should call it IUseCase, the List of the name is not necessary, what happened in the future instead to use a list you use a map, Would this method make sense in that scenario?

File: lib/src/presentation/bloc/bloc.dart [hunk 18]
```diff
@@ -0,0 +1,5 @@
+abstract class Bloc {
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/bloc.dart]: This is a IBloc right?

File: lib/src/presentation/view/movie_list_home.dart [hunk 19]
```diff
@@ -1,53 +1,99 @@
...
+        pageTitle: _movieEndpoint,
+      ),
+      drawer: Drawer(
+        child: ListView(
+          children: [
+            DrawerHeader(child: Text('Movies')),
+            ListTile(
+              title: const Text('Popular'),
+              key: Key('PopularDrawer'),
+              onTap: () => _onDrawerItemTap(MovieEndpoint.POPULAR),
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/view/movie_list_home.dart]: You can wrapped title, key and endpoint on the same enum class

Comment (FranciscoGaraventa) [reply lib/src/presentation/view/movie_list_home.dart]: You can have something like this:
```
enum MovieEndpoint {
      popular('Popular', 'PopularDrawer');
      
      const MovieEndpoint(this.title, this.key);
      
      final String title;
      final String key;
}
```

File: lib/src/data/datasource/remote/api_service.dart (hunk 1 above)
Comment (evibart) [reply lib/src/data/datasource/remote/api_service.dart]: fixed

File: lib/src/data/datasource/remote/api_service.dart (hunk 2 above)
Comment (evibart) [reply lib/src/data/datasource/remote/api_service.dart]: fixed

File: lib/src/data/datasource/remote/api_service.dart (hunk 3 above)
Comment (evibart) [reply lib/src/data/datasource/remote/api_service.dart]: fixed

File: lib/src/data/datasource/remote/api_service.dart (hunk 4 above)
Comment (evibart) [reply lib/src/data/datasource/remote/api_service.dart]: fixed

File: lib/src/data/datasource/remote/api_service.dart (hunk 5 above)
Comment (evibart) [reply lib/src/data/datasource/remote/api_service.dart]: fixed

File: lib/src/data/repository/movie_remote_repository.dart (hunk 6 above)
Comment (evibart) [reply lib/src/data/repository/movie_remote_repository.dart]: fixed

File: lib/src/presentation/bloc/bloc_movie_list.dart (hunk 7 above)
Comment (evibart) [reply lib/src/presentation/bloc/bloc_movie_list.dart]: fixed

File: lib/src/presentation/view/movie_list_home.dart (hunk 8 above)
Comment (evibart) [reply lib/src/presentation/view/movie_list_home.dart]: fixed

File: lib/my_app.dart (hunk 9 above)
Comment (evibart) [reply lib/my_app.dart]: fixed

File: lib/src/data/datasource/mappers/genre_mapper.dart (hunk 10 above)
Comment (evibart) [reply lib/src/data/datasource/mappers/genre_mapper.dart]: fixed

File: lib/src/data/datasource/mappers/movie_mapper.dart (hunk 11 above)
Comment (evibart) [reply lib/src/data/datasource/mappers/movie_mapper.dart]: fixed

File: lib/src/data/repository/genre_remote_repository.dart (hunk 12 above)
Comment (evibart) [reply lib/src/data/repository/genre_remote_repository.dart]: fixed

File: lib/src/data/repository/movie_remote_repository.dart (hunk 13 above)
Comment (evibart) [reply lib/src/data/repository/movie_remote_repository.dart]: fixed

File: lib/src/data/repository/movie_remote_repository.dart (hunk 14 above)
Comment (evibart) [reply lib/src/data/repository/movie_remote_repository.dart:7]: fixed

File: lib/src/domain/repository/movie_repository_i.dart (hunk 15 above)
Comment (evibart) [reply lib/src/domain/repository/movie_repository_i.dart]: fixed

File: lib/src/domain/usecase/implementation/up_coming_movies_usecase.dart (hunk 16 above)
Comment (evibart) [reply lib/src/domain/usecase/implementation/up_coming_movies_usecase.dart]: fixed

File: lib/src/domain/usecase/list_usecase_i.dart (hunk 17 above)
Comment (evibart) [reply lib/src/domain/usecase/list_usecase_i.dart]: fixed

File: lib/src/presentation/bloc/bloc.dart (hunk 18 above)
Comment (evibart) [reply lib/src/presentation/bloc/bloc.dart]: fixed

File: lib/src/presentation/view/movie_list_home.dart (hunk 19 above)
Comment (evibart) [reply lib/src/presentation/view/movie_list_home.dart]: fixed

File: lib/src/domain/entity/movie_entity.dart [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+        hasVideo = movieModel.hasVideo,
+        voteCount = movieModel.voteCount,
+        isAdult = movieModel.isAdult;
+
+  String get movieReleaseDate =>
+      '${releaseDate.day}/${releaseDate.month}/${releaseDate.year}';
+
+  String get posterUrl => imageBaseUrl + poster;
+
+  String get backdropUrl => imageBaseUrl + backdrop;
```
Comment (FranciscoGaraventa) [inline lib/src/domain/entity/movie_entity.dart]: You can use string interpolation here

File: lib/src/domain/repository/movie_repository_i.dart [hunk 21]
```diff
@@ -0,0 +1,12 @@
...
+import '../entity/movie_entity.dart';
+
+abstract class IMovieRepository<T> {
+  EitherResponse<List<MovieEntity>> loadMovies({required MovieEndpoint movieEndpoint});
+
+
+
+}
+
+
```
Comment (FranciscoGaraventa) [inline lib/src/domain/repository/movie_repository_i.dart]: format

**Review by MarcosCGlobant — APPROVED** [review]

**Review by LeonardoMolleker — APPROVED** [review]

File: lib/src/domain/repository/movie_repository_i.dart (hunk 21 above)
Comment (evibart) [reply lib/src/domain/repository/movie_repository_i.dart]: fixed

File: lib/src/domain/entity/movie_entity.dart (hunk 20 above)
Comment (evibart) [reply lib/src/domain/entity/movie_entity.dart]: fixed

File: lib/src/domain/entity/movie_entity.dart [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+    required this.poster,
+    required this.backdrop,
+    required this.originalLanguage,
+    required this.popularity,
+    required this.hasVideo,
+    required this.voteCount,
+    required this.isAdult,
+  });
+
+  MovieEntity.fromMovieModel(MovieModel movieModel)
```
Comment (FranciscoGaraventa) [inline lib/src/domain/entity/movie_entity.dart]: Maybe you can create a factory method here like
`factory MovieEntity.fromModel({required MovieModel model}) => MovieEntity(...)`

**Review by FranciscoGaraventa — APPROVED** [review]
Approve with minor comment

**Review by MarcosCGlobant — APPROVED** [review]

**Pull request merged**